Document and refactor goto check

xbauch opened the pull request:
If you go commit-by-commit it's not so horrible. There is not functional change only a bit of rewording.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

petr-bauch committed:
Separate checking address-of expressions

from check_rec into stand-alone function. Getting rid of the flag.

petr-bauch committed:
Clean-up and document check_rec_address

Using the checks inside to_expr methods.

petr-bauch committed:
Separate checking logical operators

into a stand-alone function.

petr-bauch committed:
Clean-up and document check_rec_logical_op

Using invariants instead of exceptions.

xbauch commented on src/analyses/goto_check.cpp:
```diff
@@ -1515,30 +1521,25 @@ void goto_checkt::check_rec_logical_op(const exprt &expr, guardt &guard)
-    throw msg;
-  }
+  INVARIANT(
+    !if_expr.cond().is_boolean(),
```
Should have been negated: sorry. Will be fixed.

tautschnig commented on src/analyses/goto_check.cpp:
```diff
@@ -102,6 +102,8 @@ class goto_checkt
   ///   operands (or their negations) for recursively calls)
   void check_rec_logical_op(const exprt &expr, guardt &guard);
 
+  void check_rec_if(const exprt &expr, guardt &guard);
```
I think this should just take an `if_exprt` instead of doing the check within the method.

xbauch replied:
Done.

tautschnig commented on src/analyses/goto_check.cpp:
```diff
@@ -110,6 +110,8 @@ class goto_checkt
   ///   the) if-condition for recursively calls)
   void check_rec_if(const exprt &expr, guardt &guard);
 
+  void check_rec_member(const exprt &expr, guardt &guard);
```
Should just take a `member_exprt`

xbauch replied:
Done.

tautschnig commented on src/analyses/goto_check.cpp:
```diff
@@ -1484,24 +1491,21 @@ void goto_checkt::check_rec_address(const exprt &expr, guardt &guard)
-      guard.add(not_exprt(op));
-    else
-      guard.add(op);
+    guard.add(expr.id() == ID_or ? not_exprt(op) : op);
```
If you are touching this, I'd suggest to use `boolean_negate(op)` instead of `not_exprt(op)`

xbauch replied:
Done.

tautschnig commented on src/analyses/goto_check.cpp:
```diff
@@ -1569,46 +1618,8 @@ void goto_checkt::check_rec(const exprt &expr, guardt &guard)
-      return;
-    }
+    check_rec_member(expr, guard);
+    return;
```
This `return` doesn't seem correct to me. I think you should remove it, and also remove the final `forall_operands` from `check_rec_member`.

xbauch replied:
The return is there only to prevent the `forall_operands` loop in case the `check_rec_member` was existed via an explicit `return`. But it seems to be only an optimisation so I did what you asked for.

xbauch replied:
I take it back. Avoiding the recursive call on the member expression operands may be just an optimisation but one of our regression tests (`Malloc23`) requires exactly 4 failing assertions. If we change this section so that the recursion is always called, this test fails.

tautschnig replied:
Which ... might be a bugfix? Does it generate more assertions or is it missing some?

xbauch replied:
Doesn't seem so:
```
struct A
{
  int x;
  int y;
};

int main()
{
  struct A *p=malloc(sizeof(int));
  p->x=0; // OK
  p->y=0; // out of bounds
..
```
Unless we skip the recursive call, we would generate a pointer check for `p->ok=0` which, on it's own, would fail.

tautschnig replied:
But still the behaviour before and after isn't the same: previously, we only returned in _some_ cases, now we always return early.

xbauch replied:
Yes, but since I included the recursive call on sub-expressions in `rec_check_member` the only thing we skip by returning early is the remaining case analyses none of which can succeed (since we know that `expr.id()==ID_member`).

tautschnig replied:
Fair enough, I was just a bit worried that one day someone would add a case. But I suppose testing (or rather: failing tests) should catch that. So it seems fine as is.

xbauch replied:
I thought about it a bit and tried having the `check_rec_member` return a Boolean indicating if the subsequent analysis should be attempted or not: https://github.com/diffblue/cbmc/pull/4611/commits/d6f7f05f2f6c0b86a96dd7a36f132dd1e1b70264.

tautschnig commented on src/analyses/goto_check.cpp:
```diff
@@ -120,6 +120,18 @@ class goto_checkt
+  ///   NaN (for floating point numbers).
+  /// \param expr: the expression to be checked
+  /// \param guard: the condition for the check (unmodified here)
+  void check_rec_div(const exprt &expr, guardt &guard);
```
Should take a `div_exprt`

xbauch replied:
Done.

tautschnig commented on src/analyses/goto_check.cpp:
```diff
@@ -1473,24 +1473,19 @@ void goto_checkt::add_guarded_claim(
-    guarded_expr = implies_exprt(guard.as_expr(), std::move(simplified_expr));
+  exprt guarded_expr =
+    guard.is_true()
+      ? simplified_expr
```
`std::move(simplified_expr)`

xbauch replied:
Done.

tautschnig commented on src/analyses/goto_check.cpp:
```diff
@@ -1489,20 +1489,16 @@ void goto_checkt::add_guarded_claim(
+      enable_assert_to_assume ? goto_programt::make_assumption(
+                                  std::move(guarded_expr), source_location)
+                              : goto_programt::make_assertion(
+                                  std::move(guarded_expr), source_location));
```
There is another instance of this pattern (around line 1800) - just search for "ASSUME". Could you please fix this as well?

xbauch replied:
Added.

tautschnig commented on src/analyses/goto_check.cpp:
```diff
@@ -275,21 +275,14 @@ void goto_checkt::invalidate(const exprt &lhs)
+        it = assertions.erase(it);
+      else
+        ++it;
     }
```
What about using `std::remove_if` (until we can use `erase_if` with C++20)?

xbauch replied:
Unfortunately, `std::remove_if` requires the iterator to be `MoveAssignable`. Will have to wait for this one.

tautschnig replied:
Ah, ok, too bad.

tautschnig reviewed (approved):
This does need cleanup as indicated in the comments, but is overall a good refactoring.

peterschrammel commented on src/analyses/goto_check.cpp:
```diff
@@ -174,11 +174,19 @@ class goto_checkt
+  /// \param src_expr: the original expression to be included in the comment
+  /// \param guard: the condition under which the asserted expression should be
+  ///   taken into account
   void add_guarded_claim(
```
While we are here, can we rename `claim` to `property` since we have replaced that terminology in other places in the codebase?

xbauch replied:
Done.

peterschrammel commented on src/analyses/goto_check.cpp:
```diff
@@ -194,6 +199,10 @@ class goto_checkt
   typedef std::set<exprt> assertionst;
   assertionst assertions;
 
+  /// Remove all assertions containing they symbol in \p lhs as well as all
```
they -> the

xbauch replied:
Done.

peterschrammel reviewed (approved):
lgtm modulo squashing promised

petr-bauch committed:
Refactor, rename, and document add_guarded_claim

into add_guarded_property.

petr-bauch committed:
Separate and refactor checking if

Using invariants, to-conversion and a bit of clang-formatting.

petr-bauch committed:
Separate and refactor checking member

Only clang-format, inline expr-construction, and for-each loop.

petr-bauch committed:
Separate and document checking arithmetic ops

Into check_rec_div and check_rec_arithmetic_op.

petr-bauch committed:
Document and refactor invalidate

Using the initializer-list constructor for sets and and standard idiom for
erase_if.

petr-bauch committed:
Use make_assumption/assertion wrappers

instead of instructiont constructor.

petr-bauch committed:
Use built-in methods

to negate boolean expression and to get address_of expression.

petr-bauch committed:
Add documentation

to collect_allocations, check_rec, and check.

petr-bauch committed:
Clang-formatting

for the bits we touched in this PR.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: d62fe0ae386de0e75515b7208b55e6c47974e5de).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112345526
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b5060ee0fca3fba80d6c977bbf052e2c3a2ada66).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112346598
